feat: benchmark existing runtime binary and select policy, presets

chungquantin opened the pull request:
<img width="1420" alt="Screenshot 2025-02-18 at 21 18 09" src="https://github.com/user-attachments/assets/5ba3e2b5-94c5-4876-8dbc-7e5e661fad45" />


# Scenarios
Context: The `pop bench pallet` command is run at the root directory of the parachain project

## Stage 1: Locating and building a `runtime` binary

- Scenario 1: If `runtime_path` (Path to the runtime, hard-coded to "runtime" folder) is not a Rust project, list all directories in the `runtime_path` (similar to `pop-node` which has multiple runtimes inside the `runtime` folder). After user select the runtime, we check if the runtime binary exists or not. If not, we build the runtime with `runtime-benchmarks` feature. 

```
list all directories -> select runtime -> check if binary exists -> build runtime with `runtime-benchmarks` feature.
```

- Scenario 2: If `runtime_path` is a Rust project, located inside the `runtime` folder, which means it is a runtime codebase (Note: There could be a chance that this is a workspace?). We will throw `No runtime found` if the workspace is not valid. We check if the runtime binary exists or not. If not, we build the runtime with `runtime-benchmarks` feature. 

```
check if binary exists -> build runtime with `runtime-benchmarks` feature.
```

- Scenario 3: If no `runtime` found, we throws `No runtime found.` (See diagram). 

## Stage 2: Select genesis builder policy and input the preset. 

After the runtime binary is ready, we prompts user to select the genesis builder (There are two options currently: `none` and `runtime`, [more details can be found here](https://github.com/paritytech/polkadot-sdk/issues/5694)). As we only support the runtime, no chain spec is provided so other genesis builder policies are not valid.

- `none` is selected: We do nothing.
- `runtime` is selected: [The runtime must be configured with `sp-genesis-builder`](https://github.com/paritytech/polkadot-sdk/pull/7476/files#diff-43828bcd5e38b4c5384e0968bca08c018b9f02edda5830bb500a91d817ee5c69)). We prompt the user to select the available genesis builder presets (presets will be registered on the runtime using `BuildGenesisConfig` runtime API). 

By calling [sc-chain-spec::GenesisConfigBuilderRuntimeCaller](https://docs.rs/sc-chain-spec/latest/sc_chain_spec/struct.GenesisConfigBuilderRuntimeCaller.html) `get_preset_names` method, we list all genesis builder presets and expect the user to select one of them. If there is no builder preset found, the CLI throws `No preset found` (can be reproduced with `pop-node`). 

- If `--runtime=` is provided, we don't locate the runtime. 
- If `--genesis-builder=` is provided, we don't prompt user to select the genesis builder policy.
- If `--genesis-config-preset=` is provided, we don't prompt user to input the genesis config preset.

# How to test the PR? 

## Stage 1 | Scenario 1: `runtime` folder is the runtime. 
- Go to `./base-parachain` or any parachain template. 
- Run `pop bench pallet --pallet=pallet_timestamp --extrinsic=` (No runtime provided)  

## Stage 1 | Scenario 2: `runtime` contains multiple runtime folders. 
- Go to `./pop-node` or any parachain that has multiple runtimes
- Run `pop bench pallet --pallet=pallet_timestamp --extrinsic=` (No runtime provided)  

## Stage 2
### `genesis-builder = runtime` with available presets.
- Go to `./base-parachain` or any parachain template. 
- Run `pop bench pallet --pallet=pallet_timestamp --extrinsic=` (No runtime provided)  
- You will be prompted to select the preset. 

![image](https://github.com/user-attachments/assets/7fc61a01-fcfa-4293-a1bb-6a2ed3568d09)

### `genesis-builder = runtime` with unavailable presets.
- Go to `./pop-node` or any parachain template. 
- Run `pop bench pallet --pallet=pallet_timestamp --extrinsic=` (No runtime provided)  
- Err thrown because there is no available genesis builder presets. 

<img width="100%" alt="Screenshot 2025-02-21 at 09 47 01" src="https://github.com/user-attachments/assets/5c1c82df-38cd-4092-ac4a-84657826cf98" />

# Changelog
- Select `runtime`, genesis builder policy and genesis config preset. 
- Add new method `build_project` to build any Rust projects with features and target, required for the tests of this PR. The method is shared with `build_parachain`. 
- Add new method `build_binary_path` to locate the binary path of the provided Rust project. The method is shared with `binary_path` method. 
- Implement new feature to automatically locate existing `<runtime-name>.wasm` to benchmark. The path of the runtime WASM blob follows: 

```
./target/<PROFILE>/wbuild/<name-of-runtime>/<name_of_runtime>.wasm
```

- Add unit tests for new methods. To test the `build_project` method for the runtime, we build the mock project with `wasm32-unknown-unknown` as target and locates the `.wasm` file. 

[sc-2931]

AlexD10S commented on crates/pop-cli/src/commands/bench/mod.rs:
```diff
@@ -72,6 +78,25 @@ impl Command {
+) -> anyhow::Result<PathBuf> {
+	let cwd = current_dir().unwrap_or(PathBuf::from("./"));
+	let target_path = mode.target_directory(&cwd).join("wbuild");
+	let project_path = cwd.join("runtime");
```
It works when I try it in the templates, but in `pop-node`, the hardcoded name doesn't work:
![Captura de pantalla 2025-02-15 a las 9 03 33](https://github.com/user-attachments/assets/08d9c713-65c7-4f93-8111-f4836d4ba962)
I'm not sure of the best approach here, can we retrieve the folder name somehow?

chungquantin replied:
Resolved.

<img width="640" alt="Screenshot 2025-02-17 at 13 06 48" src="https://github.com/user-attachments/assets/b8646e70-f83d-463f-be1b-b7d622640505" />

AlexD10S reviewed (changes requested):
The code looks very clean and well-structured, great job!

However, I encountered a couple of issues during manual testing:
- If the node has already been built without the `--runtime-benchmarks` feature,  it fails to build again because it detects the existing runtime. It throws an error (from benchmarking-cli), yet still displays a success message:
```sh
⚙  Benchmarking and generating weight file....
│  
[2025-02-15T07:41:13Z INFO  pallet_collator_selection::pallet] assembling new collators for new session 0 at #0
[2025-02-15T07:41:13Z INFO  pallet_collator_selection::pallet] assembling new collators for new session 1 at #0
[2025-02-15T07:41:13Z INFO  polkadot_sdk_frame::benchmark::pallet] Loading WASM from file
└  Failed to run benchmarking: Invalid input: Could not call runtime API to Did not find the benchmarking metadata. This could mean that you either did not build the node correctly with the `--features runtime-benchmarks` flag, or the chain spec that you are using was not created by a node that was compiled with the flag: Other: Exported method Benchmark_benchmark_metadata is not found
└  Benchmark completed successfully!
```
Probably we can't verify whether the runtime has been build with `--runtime-benchmarks` at this point right? It would be ideal if we could, but if not, we should at least prevent the misleading success message. I can see here, it shows a message with the error but continue the execution: https://github.com/r0gue-io/pop-cli/blob/chugnquantin/feat-build_runtime/crates/pop-cli/src/commands/bench/mod.rs#L68

- The second issue relates to locating the WASM file, see my comment below.



Out of scope for this PR, but once the benchmarking logic is merged into main, we could consider adding a flag in pop build to handle building with the `--runtime-benchmarks ` feature.

chungquantin commented on crates/pop-cli/src/commands/mod.rs:
```diff
@@ -5,6 +5,7 @@ use clap::Subcommand;
 use pop_common::templates::Template;
 use serde_json::{json, Value};
 
+#[cfg(feature = "parachain")]
```
Resolved by rebasing to the main PR

chungquantin commented on crates/pop-cli/src/commands/bench/mod.rs:
```diff
@@ -54,66 +56,122 @@ impl Command {
-
-		if let Some(output_path) = cmd.output {
-			console::Term::stderr().clear_last_lines(1)?;
-			cli.info(format!("Weight file is generated to {}", output_path.as_path().display()))?;
```
I don't think this is necessary so removed

chungquantin commented on crates/pop-cli/src/commands/bench/mod.rs:
```diff
@@ -1,16 +1,23 @@
 /// Arguments for bencharmking a project.
 #[derive(Args)]
-#[command(args_conflicts_with_subcommands = true)]
+#[command(args_conflicts_with_subcommands = true, ignore_errors = true)]
```
Bypass the argument parsing checks. This is required for `runtime`, `pallet` and `extrinsic`

AlexD10S reviewed (commented):
Tests need to be fixed first, but here are some comments I found.

- A important thing to fix is that I can’t currently run:
```
pop bench pallet --help
```
- Additionally, running:
```
pop bench pallet -pallet=pallet_timestamp --extrinsic=
```
in either the `pop-cli` repo or a contract repo causes a panic. While this isn't critical for a CLI, it would be ideal to handle it with a proper error message.
I think this is due to the assumption mentioned in the PR description:
_Scenario 2: If runtime_path is a Rust project, meaning it is a runtime codebase._

- How can I test if the genesis builder policy: `runtime` working? When I tried, I got the `Error: Invalid genesis builder option: runtime.`

By the way, I loved the diagram in the description!

chungquantin commented:
> Tests need to be fixed first, but here are some comments I found.
> 
> * A important thing to fix is that I can’t currently run:
> 
> ```
> pop bench pallet --help
> ```
> 
> * Additionally, running:
> 
> ```
> pop bench pallet -pallet=pallet_timestamp --extrinsic=
> ```
> 
> in either the `pop-cli` repo or a contract repo causes a panic. While this isn't critical for a CLI, it would be ideal to handle it with a proper error message. I think this is due to the assumption mentioned in the PR description: _Scenario 2: If runtime_path is a Rust project, meaning it is a runtime codebase._
> 
> * How can I test if the genesis builder policy: `runtime` working? When I tried, I got the `Error: Invalid genesis builder option: runtime.`
> 
> By the way, I loved the diagram in the description!

- I have removed the `ignore_errors` for this PR and the issue with the `--help` command is resolved. 
- About panicking if the path is detected as a Rust project,  I have added code to check if the project is located in `runtime` folder first. One nice-to-have approach that we can implement in the future is asking user to input the path to the runtime project. 
- `runtime` genesis builder policy is fixed. This works with `base-parachain` but not `pop-node`. I expect parachains that are not configured with genesis builder properly on the runtime should choose `none` for the policy instead of `runtime`.

AlexD10S reviewed (commented):
It looks good! I just left some questions for clarification. 
Also, I noticed that the `build_runtime_works` test is failing in CI. I tried rerunning it, but it still fails, even though it works locally. Any thoughts?

chungquantin commented on tests/runtimes/base_parachain.wasm:
I moved it to `~/tests/runtimes` because tests in `pop-parachains` and `pop-cli` need this.

chungquantin commented on crates/pop-parachains/src/bench.rs:
```diff
@@ -1,17 +1,108 @@
+/// * `binary_path` - Path to the runtime WASM binary.
+/// * `preset` - Optional ID of the genesis config preset. If not provided, it checks the default
+///   preset.
+pub fn check_preset(binary_path: &PathBuf, preset: Option<&String>) -> anyhow::Result<()> {
```
Updated to only display the preset selection stage if there are available presets. Default to `none` for the policy if no presets found
